Mathlib/RingTheory/Valuation/ValuativeRel.lean

```diff
@@ -0,0 +1,587 @@
+  not_rel_one_zero := by simp
+
+lemma isEquiv {Γ₁ Γ₂ : Type*}
+    [LinearOrderedCommMonoidWithZero Γ₁]
+    [LinearOrderedCommMonoidWithZero Γ₂]
+    (v₁ : Valuation R Γ₁)
+    (v₂ : Valuation R Γ₂)
+    [v₁.Compatible] [v₂.Compatible] :
+    v₁.IsEquiv v₂ := by
+  intro x y
+  simp_rw [← Valuation.Compatible.rel_iff_le]
+
+variable (R) in
+/-- An alias for endowing a ring with a preorder defined as the valuative relation. -/
+def WithPreorder := R
+
+/-- The ring instance on `WithPreorder R` arising from the ring structure on `R`. -/
+instance : CommRing (WithPreorder R) := inferInstanceAs (CommRing R)
+
+/-- The preorder on `WithPreorder R` arising from the valuative relation on `R`. -/
+instance : Preorder (WithPreorder R) where
+  le (x y : R) := x ≤ᵥ y
+  le_refl _ := rel_refl _
+  le_trans _ _ _ := rel_trans
+
+/-- The valutaive relation on `WithPreorder R` arising from the valuative relation on `R`.
+This is defined as the preorder itself. -/
+instance : ValuativeRel (WithPreorder R) where
+  rel := (· ≤ ·)
+  rel_total := rel_total (R := R)
+  rel_trans := rel_trans (R := R)
+  rel_add := rel_add (R := R)
+  rel_mul_right := rel_mul_right (R := R)
+  rel_mul_cancel := rel_mul_cancel (R := R)
+  not_rel_one_zero := not_rel_one_zero (R := R)
+
+instance : ValuativePreorder (WithPreorder R) where
+  rel_iff_le _ _ := Iff.rfl
+
+variable (R) in
+/-- The support of the valuation on `R`. -/
+def supp : Ideal R where
+  carrier := { x | x ≤ᵥ 0 }
+  add_mem' ha hb := rel_add ha hb
+  zero_mem' := rel_refl _
+  smul_mem' x _ h := by simpa using rel_mul_left _ h
+
+@[simp]
+lemma supp_def (x : R) : x ∈ supp R ↔ x ≤ᵥ 0 := Iff.refl _
+
+lemma supp_eq_valuation_supp : supp R = (valuation R).supp := by
+  ext x
+  constructor
+  · intro h
+    simp only [supp_def, Valuation.mem_supp_iff] at h ⊢
+    apply ValGroupWithZero.sound
+    · simpa
+    · simp
+  · intro h
+    have := ValGroupWithZero.exact h
+    simpa using this.left
+
+instance : (supp R).IsPrime := by
+  rw [supp_eq_valuation_supp]
+  infer_instance
+
+open NNReal in variable (R) in
+/-- An auxiliary structure used to define `IsRankOne`. -/
+structure RankOneStruct where
+  /-- The embedding of the value group-with-zero into the nonnegative reals. -/
+  emb : ValGroupWithZero R →*₀ ℝ≥0
+  strictMono : StrictMono emb
+  nontrivial : ∃ γ : ValGroupWithZero R, emb γ ≠ 0 ∧ emb γ ≠ 1
+
+variable (R) in
+/-- We say that a ring with a valuative relation is of rank one if
+there exists a strictly monotone embedding of the "canonical" value group-with-zero into
+the nonnegative reals, and the image of this embedding contains some element different
+from `0` and `1`. -/
+class IsRankOne where
+  nonempty : Nonempty (RankOneStruct R)
+
+variable (R) in
+/-- A ring with a valuative relation is discrete if its value group-with-zero
+has a maximal element `< 1`. -/
+class IsDiscrete where
+  has_maximal_element :
+    ∃ γ : ValGroupWithZero R, γ < 1 ∧ (∀ δ : ValGroupWithZero R, δ < 1 → δ ≤ γ)
```

Comment:
Wait -- are we happy for trivial valuations (value group is trivial) to be `Discrete`? Are we happy for rank 2 valuations to be discrete?